Sources/NIOCore/AsyncChannel/AsyncChannelOutboundWriterHandler.swift

```diff
@@ -0,0 +1,180 @@
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if swift(>=5.6)
+import DequeModule
+
+/// A ``ChannelHandler`` that is used to write the outbound portion of a NIO
+/// ``Channel`` from Swift Concurrency with back-pressure support.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+@usableFromInline
+internal final class NIOAsyncChannelOutboundWriterHandler<OutboundOut: Sendable>: ChannelDuplexHandler {
+    @usableFromInline typealias InboundIn = Any
+    @usableFromInline typealias InboundOut = Any
+    @usableFromInline typealias OutboundIn = Any
+    @usableFromInline typealias OutboundOut = OutboundOut
+
+    @usableFromInline
+    typealias Writer = NIOAsyncWriter<
+        OutboundOut,
+        NIOAsyncChannelOutboundWriterHandler<OutboundOut>.Delegate
+    >
+
+    @usableFromInline
+    typealias Sink = Writer.Sink
+
+    /// The sink of the ``NIOAsyncWriter``.
+    @usableFromInline
+    var sink: Sink?
+
+    /// The channel handler context.
+    @usableFromInline
+    var context: ChannelHandlerContext?
+
+    /// The event loop.
+    @usableFromInline
+    let eventLoop: EventLoop
+
+    /// The shared `CloseRatchet` between this handler and the inbound stream handler.
+    @usableFromInline
+    let closeRatchet: CloseRatchet
+
+    /// Indicates if the handler should do half closure.
+    @usableFromInline
+    let enableOutboundHalfClosure: Bool
+
+    @inlinable
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet,
+        enableOutboundHalfClosure: Bool
+    ) {
+        self.eventLoop = eventLoop
+        self.closeRatchet = closeRatchet
+        self.enableOutboundHalfClosure = enableOutboundHalfClosure
+    }
+
+    @inlinable
+    func _didYield(sequence: Deque<OutboundOut>) {
+        // This is always called from an async context, so we must loop-hop.
+        // Because we always loop-hop, we're always at the top of a stack frame. As this
+        // is the only source of writes for us, and as this channel handler doesn't implement
+        // func write(), we cannot possibly re-entrantly write. That means we can skip many of the
+        // awkward re-entrancy protections NIO usually requires, and can safely just do an iterative
+        // write.
+        self.eventLoop.preconditionInEventLoop()
+        guard let context = self.context else {
+            // Already removed from the channel by now, we can stop.
+            return
+        }
+
+        self._doOutboundWrites(context: context, writes: sequence)
+    }
+
+    @inlinable
+    func _didTerminate(error: Error?) {
+        self.eventLoop.preconditionInEventLoop()
+
+        switch self.closeRatchet.closeWrite() {
+        case .nothing:
+            if self.enableOutboundHalfClosure {
+                self.context?.close(mode: .output, promise: nil)
+            }
```

Comment:
This seems wrong (or at the very least is misleading): why are we half closing when the ratchet told us to do nothing?

Reply:
It is doing the right thing but the return of the `CloseRatchet` is confusing. I am going to clean that up.